idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,987 @@
+    extreme_jacobian_entries,
+    jacobian_cond,
+)
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+MAX_STR_LENGTH = 84
+TAB = " " * 4
+
+# TODO: Add suggested steps to cautions - how?
+
+CONFIG = ConfigDict()
+CONFIG.declare("model", ConfigValue(description="Pyomo model object to be diagnosed."))
+CONFIG.declare(
+    "absolute_tolerance",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Absolute tolerance to for variables close to bounds.",
+    ),
+)
+CONFIG.declare(
+    "relative_tolerance",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Relative tolerance to for variables close to bounds.",
+    ),
+)
+CONFIG.declare(
+    "residual_tolerance",
+    ConfigValue(
+        default=1e-5,
+        domain=float,
+        description="Absolute tolerance to use when checking constraint residuals.",
+    ),
+)
+CONFIG.declare(
+    "large_value_tolerance",
+    ConfigValue(
+        default=1e4,
+        domain=float,
+        description="Absolute tolerance for considering a value to be large.",
+    ),
+)
+CONFIG.declare(
+    "small_value_tolerance",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Absolute tolerance for considering a value to be small.",
+    ),
+)
+CONFIG.declare(
+    "zero_value_tolerance",
+    ConfigValue(
+        default=1e-8,
+        domain=float,
+        description="Absolute tolerance for considering a value to be near to zero.",
+    ),
+)
+CONFIG.declare(
+    "jacobian_large_value_caution",
+    ConfigValue(
+        default=1e4,
+        domain=float,
+        description="Tolerance for raising a caution for large Jacobian values.",
+    ),
+)
+CONFIG.declare(
+    "jacobian_large_value_warning",
+    ConfigValue(
+        default=1e8,
+        domain=float,
+        description="Tolerance for raising a warning for large Jacobian values.",
+    ),
+)
+CONFIG.declare(
+    "jacobian_small_value_caution",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Tolerance for raising a caution for small Jacobian values.",
+    ),
+)
+CONFIG.declare(
+    "jacobian_small_value_warning",
+    ConfigValue(
+        default=1e-8,
+        domain=float,
+        description="Tolerance for raising a warning for small Jacobian values.",
+    ),
+)
+
+
+@document_kwargs_from_configdict(CONFIG)
+class DiagnosticsToolbox:
+    """
+    The IDAES Model DiagnosticsToolbox.
+
+    To get started:
+
+      1. Create an instance of your model (this does not need to be initialized yet).
+      2. Fix variables until you have 0 degrees of freedom. Many of these tools presume
+         a square model, and a square model should always be the foundation of any more
+         advanced model.
+      3. Create an instance of the DiagnosticsToolbox and provide the model to debug as
+         the model argument.
+      4. Call the report_structural_issues() method.
```

Comment:
Can you format "report_structural_issues()" as code? I.e. with double back-ticks